### ISSUE TITLE
When using XamlAnimatedGif to display online GIF images, the CPU resources are over-utilized.

### ISSUE THREAD

**Issue by qian-o** [XamlAnimatedGif/XamlAnimatedGif#160]
A lag occurs when multiple network resources are played simultaneously.  
CPU : 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
RAM : 16.0 GB
GPU : RTX 3060 6G
![JS`ZMT@V0F_6JILG77JU(DL](https://user-images.githubusercontent.com/84434846/134471029-d3e771df-d0b8-4e02-a0e3-bbac351507fb.png)

URI : http://47.96.133.119/api/Attachment/GetAttachments/eIfN3z27yS_xNPXPhTRee.gif

**Comment by thomaslevesque** [comment]
Hi @qian-o,

Yes, I'm aware that this library has performance issues. It's a tradeoff between memory usage and CPU usage. This library was designed to reduce memory usage compared to [WpfAnimatedGif](https://github.com/XamlAnimatedGif/WpfAnimatedGif/), but it means doing much more work on the CPU.

The fact that it's a network resource shouldn't have much impact; the file is downloaded locally before playing it.

I have plans to improve performance, but
1. displaying multiple GIFs, especially large ones, will always be costly in terms of CPU usage
2. I honestly have no idea when I'll get around to make those improvements. I have little free time, so probably not in the short term.

**Comment by qian-o** [comment]
I tried to compare the performance of web resource and local resource. Under the same image, web resource would occupy more CPU, which is 3 times of that occupied by local resource.
Besides, playing web resource in the XamlAnimatedGif.Demo you gave would also occupy a significant share of CPU.
Local resource:
![image](https://user-images.githubusercontent.com/84434846/134483175-9224bed9-3af2-43f2-8184-92fe36fced75.png)

Web resource:
![image](https://user-images.githubusercontent.com/84434846/134483354-8b99aec7-56ab-4aa7-98d3-a729f7e03b80.png)

**Comment by thomaslevesque** [comment]
@qian-o I'm not seeing this on my machine. The CPU load is about the same when loading the image from a URL or from a local file (but when loading from a URL for the first time there's a delay before the image shows up, due to the image is being downloaded)

**Comment by qian-o** [comment]
After repeated tests, I found that when the GIF generation operation is changed to resource and played in the form of a relative path, there will be very little resource occupation.
If it is played in the form of absolute path such as local resource or network resource, it will have extremely high resource occupation.

Resource：
![image](https://user-images.githubusercontent.com/84434846/134531487-f974601e-c127-467c-a831-debacc1f3d8d.png)

Absolute path：
![image](https://user-images.githubusercontent.com/84434846/134531711-5d85ab38-1c80-479d-89b9-81bc0ec13e4a.png)

**Comment by thomaslevesque** [comment]
I can't reproduce your results. I tried to compare CPU usage between absolute file path and relative path to a resource, but it doesn't show any significant difference for me. Anyway, the Windows Task Manager isn't a reliable tool for this kind of thing. Many other things can cause spikes in CPU usage.
If your results can't be reproduced on another machine, I'm afraid there's not much I can do...

**Comment by qian-o** [comment]
Thank you for your reply. I will continue to test in different environments.

**Comment by qian-o** [comment]
I tried to write the bytes read by FileStream into MemoryStream to do buffer, which could effectively improve the performance. 
FileStream：
![image](https://user-images.githubusercontent.com/84434846/134610722-ec30cf4e-9dd2-467f-a738-83bc1873a797.png)

MemoryStream：
![image](https://user-images.githubusercontent.com/84434846/134610776-bf4eccc5-2e19-4e72-a8a9-c822609bef4f.png)

I think the FileStream used to read files will affect the performance significantly and MemoryStream should be used to do buffer.

**Comment by thomaslevesque** [comment]
Yes, loading the file in memory could help, however it would also use more RAM, which is an issue for large GIFs...

**Comment by qian-o** [comment]
I hope that you will weigh the frequency of using RAM or IO. If there are too many GIF files, they will have a significant impact on the performance.

**Comment by thomaslevesque** [comment]
Maybe I could make it optional on a case-by-case basis

**Comment by qian-o** [comment]
OK, thank you for your reply.

**Comment by thomaslevesque** [comment]
Actually, I don't think I will implement this in the library itself, because it can easily be done by the user, and handling this to the library would just add unnecessary complexity.
You can just copy the `FileStream` to a `MemoryStream` and pass the `MemoryStream` to `SetSourceStream`, like you did in your tests.

**Comment by thomaslevesque** [comment]
Actually, this might actually be useful. While it's true that the user could do the memory caching themselves, it means being unable to set the source as a URI. I'll try to add an option to load images in memory in a future version.
